### PR TITLE
Simplify side conditions

### LINKED ISSUE #1416: Include the side condition when evaluating functions in the side condition
kevm: https://github.com/kframework/evm-semantics/tree/investigate

```
make clean
make defn

kompile --debug --main-module ETHEREUM-SIMULATION --backend haskell --hook-namespaces KRYPTO --syntax-module ETHEREUM-SIMULATION .build/defn/haskell/driver.k --directory .build/defn/haskell -I .build/defn/haskell

kprove --haskell-backend-command "/home/zy/workspace/kore/.build/kore/bin/kore-repl --repl-script /home/zy/workspace/kore/dist/kast.kscript " ~/workspace/evm-semantics/tests/specs/erc20/ds/balanceOf-spec.k -d ~/workspace/evm-semantics/.build/defn/haskell/ --def-module VERIFICATION
```
print the configuration at step 0 yields a very complicated predicate like:
```
  ( #Not ( {
        0 <=Int OWNER andBool OWNER <=Int 115792089237316195423570985008687907853269984665640564039457584007913129639935
      #Equals
        true
      } )
    #And
      {
        #lookup ( M , keccak ( #buf ( 32 , OWNER ) +Bytes b"0000000000000000000000000000000000000000000000000000000000000001" ) )
      #Equals
        #lookup ( M , keccak ( intList2ByteArray ( OWNER  1  .IntList ) ) )
      }
  #Implies
    {
      true
    #Equals
      0 <=Int ACCT_ID andBool ACCT_ID <=Int 1461501637330902918203684832716283019655932542975 andBool ( 0 <=Int CALLER_ID andBool CALLER_ID <=Int 1461501637330902918203684832716283019655932542975 ) andBool ( 0 <=Int ORIGIN_ID andBool ORIGIN_ID <=Int 1461501637330902918203684832716283019655932542975 ) andBool ( 0 <=Int CALL_DEPTH andBool CALL_DEPTH <Int 1024 ) andBool ( 0 <=Int OWNER andBool OWNER <=Int 1461501637330902918203684832716283019655932542975 ) andBool ( 0 <=Int #lookup ( M , keccak ( intList2ByteArray ( OWNER  1  .IntList ) ) ) andBool #lookup ( M , keccak ( intList2ByteArray ( OWNER  1  .IntList ) ) ) <=Int 115792089237316195423570985008687907853269984665640564039457584007913129639935 ) andBool #isIntegerMap ( M )
    } )
#And
  ( {
      0 <=Int OWNER andBool OWNER <=Int 115792089237316195423570985008687907853269984665640564039457584007913129639935
    #Equals
      true
    }
  #Implies
    {
      true
    #Equals
      0 <=Int ACCT_ID andBool ACCT_ID <=Int 1461501637330902918203684832716283019655932542975 andBool ( 0 <=Int CALLER_ID andBool CALLER_ID <=Int 1461501637330902918203684832716283019655932542975 ) andBool ( 0 <=Int ORIGIN_ID andBool ORIGIN_ID <=Int 1461501637330902918203684832716283019655932542975 ) andBool ( 0 <=Int CALL_DEPTH andBool CALL_DEPTH <Int 1024 ) andBool ( 0 <=Int OWNER andBool OWNER <=Int 1461501637330902918203684832716283019655932542975 ) andBool ( 0 <=Int #lookup ( M , keccak ( #buf ( 32 , OWNER ) +Bytes b"0000000000000000000000000000000000000000000000000000000000000001" ) ) andBool #lookup ( M , keccak ( #buf ( 32 , OWNER ) +Bytes b"0000000000000000000000000000000000000000000000000000000000000001" ) ) <=Int 115792089237316195423570985008687907853269984665640564039457584007913129639935 ) andBool #isIntegerMap ( M )
    } )
#And
  #Ceil ( SetItem ( ACCT_ID:Int )
  _34 )
#And
  #Not ( #Not ( {
      0 <=Int OWNER andBool OWNER <=Int 115792089237316195423570985008687907853269984665640564039457584007913129639935
    #Equals
      true
    } )
  #And
    #Not ( {
      #lookup ( M , keccak ( #buf ( 32 , OWNER ) +Bytes b"0000000000000000000000000000000000000000000000000000000000000001" ) )
    #Equals
      #lookup ( M , keccak ( intList2ByteArray ( OWNER  1  .IntList ) ) )
    } ) )
```

As @virgil-serbanuta pointed out, it may be caused by evaluating
```
andBool BAL ==Int #lookup(M, #hashedLocation("Solidity", 1, OWNER))
``` 
in the top level predicate of the spec. When evaluating `#hashedLocation`, it needs to apply the rule
```
    syntax ByteArray ::= intList2ByteArray( IntList ) [function, functional]
 // ------------------------------------------------------------------------
    rule intList2ByteArray(.IntList) => .ByteArray
    rule intList2ByteArray(V VS)     => #buf(32, V) ++ intList2ByteArray(VS)
      requires #rangeUInt(256, V)
```
And applying the rule requires using the following condition in the top level predicate:
```
andBool #rangeAddress(OWNER)
```

However, when evaluating the functions in top level predicate, we don't pass in side conditions.

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes #1416

If the side condition contains multiple conditions, these changes will allow each condition to be simplified one time under the assumption that the others are true.

This PR also removes all the cases where we simplify the side condition with itself.

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/src/Kore/Step/Simplification/Pattern.hs [hunk 1]
```diff
@@ -89,7 +116,36 @@ simplify sideCondition pattern' =
...
+            ()
+    worker [] = return ()
+    worker (pred' : rest) = do
+        sideCondition <- State.get
+        let otherConds = sideCondition <> from (MultiAnd.make rest)
+        result <-
+            simplifyCondition otherConds (Condition.fromPredicate pred')
+            & lift
+        State.put (SideCondition.andCondition sideCondition result)
+        worker rest
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/Pattern.hs:151]: Let me make sure I understand this correctly: The `SideCondition` comprised of all simplified `Predicate`s is held in the `State`. At each step, we combine that with the `SideCondition` from all (other) un-simplified `Predicate`s, and use the combination to simplify the current `Predicate`. Then we update the stored `SideCondition`. Is that right?

**Review by ttuegel — COMMENTED** [review]
For a test of the `Pattern` simplifier, I recommend simplifying something with a side condition that can be evaluated by function rules. For example, `a ∧ (f(a) = b ∧ g(a) = g(b))` where the simplifier knows a function rule that says `f(a) = b requires g(a) = g(b)`.

File: kore/src/Kore/Step/Simplification/Pattern.hs (hunk 1 above)
Comment (ana-pantilie) [reply kore/src/Kore/Step/Simplification/Pattern.hs:151]: That's right.

**Comment by ttuegel** [comment]
To debug the performance regression, you can try the Kore-level profiler with speedscope: https://github.com/kframework/kore/wiki/Profiling#kore-profiling

**Comment by ana-pantilie** [comment]
Error on `test/sidecondition/remove-12-spec.k`:

```
kore-exec: [1929158] Error (ErrorException):
    Internal error: expected simplified results, but found:
        \and{_PREDICATE{}}(
            /* term: */
            /* D Sfa */ \top{_PREDICATE{}}(),
        \and{_PREDICATE{}}(
            /* predicate: */
            \and{_PREDICATE{}}(
                /* Sfc */
                \equals{SortBool{}, _PREDICATE{}}(
                    /* Fl Fn D Sfa Cl */ \dv{SortBool{}}("false"),
                    /* Fl Fn D Sfc */
                    Lbl'Unds'in'Unds'keys'LParUndsRParUnds'MAP'Unds'Bool'Unds'KItem'Unds'Map{}(
                        /* Fl Fn D Sfc */
                        /* Inj: */ inj{SortMyId{}, SortKItem{}}(
                            /* Fl Fn D Sfa */ RuleVarY:SortMyId{}
                        ),
                        /* Fl Fn D Sfa */ RuleVarMAP:SortMap{}
                    )
                ),
            \and{_PREDICATE{}}(
                /* Sfc */
                \forall{_PREDICATE{}}(
                    Rulex:SortKItem{},
                    /* Sfc */
                    \ceil{SortMap{}, _PREDICATE{}}(
                        /* Fn Spc */
                        /* InternalMap: */ Lbl'Unds'Map'Unds'{}(
                            /* element: */ Lbl'UndsPipe'-'-GT-Unds'{}(
                                /* Fl Fn D Sfc */
                                /* Inj: */ inj{SortMyId{}, SortKItem{}}(
                                    /* Fl Fn D Sfa */ RuleVarX:SortMyId{}
                                ),
                                /* Fl Fn D Sfa */ Rulex:SortKItem{}
                            ),
                            /* opaque child: */ /* Fl Fn D Sfa */
                            RuleVarMAP:SortMap{}
                        )
                    )
                ),
                /* Sfa */
                \not{_PREDICATE{}}(
                    /* Spa */
                    \equals{SortMyId{}, _PREDICATE{}}(
                        /* Fl Fn D Sfa */ RuleVarX:SortMyId{},
                        /* Fl Fn D Sfa */ RuleVarY:SortMyId{}
                    )
                )
            )),
            /* substitution: */
            \top{_PREDICATE{}}()
        ))
      while simplifying:
        /* Spa */
        \equals{SortBool{}, _PREDICATE{}}(
            /* Fn Spa */
            LblnotBool'Unds'{}(
                /* Fn Spa */
                Lbl'Unds'in'Unds'keys'LParUndsRParUnds'MAP'Unds'Bool'Unds'KItem'Unds'Map{}(
                    /* Fl Fn D Spa */
                    /* Inj: */ inj{SortMyId{}, SortKItem{}}(
                        /* Fl Fn D Sfa */ RuleVarY:SortMyId{}
                    ),
                    /* Fn Spa */
                    /* InternalMap: */ Lbl'Unds'Map'Unds'{}(
                        /* element: */ Lbl'UndsPipe'-'-GT-Unds'{}(
                            /* Fl Fn D Spa */
                            /* Inj: */ inj{SortMyId{}, SortKItem{}}(
                                /* Fl Fn D Sfa */ RuleVarX:SortMyId{}
                            ),
                            /* Fl Fn D Sfa Cli */
                            /* Inj: */ inj{SortInt{}, SortKItem{}}(
                                /* Fl Fn D Sfa Cl */ \dv{SortInt{}}("1")
                            )
                        ),
                        /* opaque child: */ /* Fl Fn D Sfa */ RuleVarMAP:SortMap{}
                    )
                )
            ),
            /* Fl Fn D Sfa Cl */ \dv{SortBool{}}("true")
        )
      with side condition:
        /* Spa */
        \and{_PREDICATE{}}(
            /* Spa */
            \equals{SortBool{}, _PREDICATE{}}(
                /* Fl Fn D Spa */
                Lbl'UndsEqlsEqls'K'Unds'{}(
                    /* Fl Fn D Spa */
                    kseq{}(
                        /* Fl Fn D Spa */
                        /* Inj: */ inj{SortMyId{}, SortKItem{}}(
                            /* Fl Fn D Sfa */ VarZ:SortMyId{}
                        ),
                        /* Fl Fn D Sfa Cl */ dotk{}()
                    ),
                    /* Fl Fn D Spa */
                    kseq{}(
                        /* Fl Fn D Spa */
                        /* Inj: */ inj{SortMyId{}, SortKItem{}}(
                            /* Fl Fn D Sfa */ VarY:SortMyId{}
                        ),
                        /* Fl Fn D Sfa Cl */ dotk{}()
                    )
                ),
                /* Fl Fn D Sfa Cl */ \dv{SortBool{}}("true")
            ),
            /* Spa */
            \equals{SortGeneratedCounterCell{}, _PREDICATE{}}(
                /* Fl Fn D Sfa */ Var'Unds'DotVar0:SortGeneratedCounterCell{},
                /* Fl Fn D Spa */
                Lbl'-LT-'generatedCounter'-GT-'{}(
                    /* Created: Kore.Step.Rule.Expand.maybeNewVariable */
                    /* Fl Fn D Sfa */
                    Var'Unds'DotVar1:SortInt{}
                )
            )
        )
    CallStack (from HasCallStack):
      error, called at src/Kore/Step/Simplification/TermLike.hs:485:6 in kore-0.39.0.0-DgV0ei696gh4fIMAGEIAtF:Kore.Step.Simplification.TermLike
```

File: kore/src/Kore/Attribute/Pattern/Simplified.hs [hunk 2]
```diff
@@ -240,6 +241,10 @@ isFullySimplified (Simplified_ Fully Unknown) = False
 isFullySimplified (Simplified_ Partly _) = False
 isFullySimplified NotSimplified = False
 
+isSimplifiedIgnoreCondition :: Simplified -> Bool
```
Comment (ttuegel) [inline kore/src/Kore/Attribute/Pattern/Simplified.hs:244]:
```suggestion
{- | Is the pattern fully simplified under some side condition?

See also: 'isSimplified', 'isFullySimplified'.

 -}
isSimplifiedIgnoreCondition :: Simplified -> Bool
```

File: kore/src/Kore/Attribute/PredicatePattern.hs [hunk 3]
```diff
@@ -97,6 +100,10 @@ isSimplified
     :: SideCondition.Representation -> PredicatePattern variable -> Bool
 isSimplified sideCondition = Simplified.isSimplified sideCondition . simplifiedAttribute
 
+isSimplifiedIgnoreCondition
```
Comment (ttuegel) [inline kore/src/Kore/Attribute/PredicatePattern.hs:103]: Please add documentation similar to `Kore.Attribute.Pattern.isSimplifiedIgnoreCondition`.

File: kore/src/Kore/Attribute/Pattern.hs [hunk 4]
```diff
@@ -156,6 +159,13 @@ isSimplified sideCondition patt@Pattern {simplified} =
     assertSimplifiedConsistency patt
     $ Simplified.isSimplified sideCondition simplified
 
+isSimplifiedIgnoreCondition
```
Comment (ttuegel) [inline kore/src/Kore/Attribute/Pattern.hs:162]: Please add documentation similar to `Kore.Attribute.Pattern.Simplified.isSimplifiedIgnoreCondition`.

File: kore/src/Kore/Internal/Predicate.hs [hunk 5]
```diff
@@ -914,6 +916,9 @@ simplifiedAttribute = Attribute.simplifiedAttribute . extractAttributes
 isSimplified :: SideCondition.Representation -> Predicate variable -> Bool
 isSimplified condition = Attribute.isSimplified condition . extractAttributes
 
+isSimplifiedIgnoreCondition :: Predicate variable -> Bool
```
Comment (ttuegel) [inline kore/src/Kore/Internal/Predicate.hs:919]: Please add documentation similar to `Kore.Attribute.Pattern.Simplified.isSimplifiedIgnoreCondition`.

File: kore/src/Kore/Internal/Substitution.hs [hunk 6]
```diff
@@ -493,6 +494,11 @@ isSimplified _ (Substitution _) = False
 isSimplified sideCondition (NormalizedSubstitution normalized) =
     all (TermLike.isSimplified sideCondition) normalized
 
+isSimplifiedIgnoreCondition :: Substitution variable -> Bool
```
Comment (ttuegel) [inline kore/src/Kore/Internal/Substitution.hs:497]: Please add documentation similar to `Kore.Attribute.Pattern.Simplified.isSimplifiedIgnoreCondition`.

File: kore/src/Kore/Internal/TermLike.hs [hunk 7]
```diff
@@ -449,6 +450,10 @@ isSimplified :: SideCondition.Representation -> TermLike variable -> Bool
 isSimplified sideCondition =
     Attribute.isSimplified sideCondition . extractAttributes
 
+isSimplifiedIgnoreCondition :: TermLike variable -> Bool
```
Comment (ttuegel) [inline kore/src/Kore/Internal/TermLike.hs:453]: Please add documentation.

**Review by ttuegel — APPROVED** [review]
Please add some documentation for the new functions, even if it is repetitive.

I also thought about how we could name the `isSimplified_` family of functions to call attention to their differences:

- `isSimplified` (that has a side condition input) stays the same.
- `isFullySimplified` becomes `isSimplifiedAnyCondition`.
- `isSimplifiedIgnoreCondition` becomes `isSimplifiedSomeCondition`.

We don't have to do that; let me know what you think.

**Pull request merged**